source/Coop.Core/Server/Services/Sync/Handlers/PeerQueueOverloadedHandler.cs

```diff
@@ -0,0 +1,114 @@
+﻿using Common;
+using Common.Messaging;
+using Common.Network;
+using Coop.Core.Server.Services.Time.Handlers;
+using Coop.Core.Server.Services.Time.Messages;
+using GameInterface.Services.GameDebug.Messages;
+using GameInterface.Services.Heroes.Enum;
+using LiteNetLib;
+using System;
+using System.Collections.Generic;
+namespace Coop.Core.Server.Services.Sync.Handlers
+{
+    internal class PeerQueueOverloadedHandler : IHandler
+    {
+        private static readonly long POLL_INTERVAL = 100;
+
+        private readonly IMessageBroker messageBroker;
+        private readonly INetwork network;
+
+        private readonly Poller poller;
+        private readonly HashSet<NetPeer> overloadedPeers;
+
+        private readonly TimeHandler timeHandler;
+
+        private TimeControlEnum originalSpeed;
+
+        public PeerQueueOverloadedHandler(
+            IMessageBroker messageBroker, 
+            INetwork network,
+            TimeHandler timeHandler
+        )
+        {
+            this.messageBroker = messageBroker;
+            this.network = network;
+            this.timeHandler = timeHandler;
+
+            overloadedPeers = new();
+            poller = new(Poll, TimeSpan.FromMilliseconds(POLL_INTERVAL));
+
+            messageBroker.Subscribe<PeerQueueOverloaded>(Handle);
+
+            poller.Start();
+        }
+
+        private void Handle(MessagePayload<PeerQueueOverloaded> payload)
+        {
+            lock (overloadedPeers)
+            {
+                if (overloadedPeers.Add(payload.What.NetPeer))
```

Comment:
Instead of locking for the whole process, it should work the same if you only lock the Add()

```suggestion
            lock (overloadedPeers)
            {
                if (overloadedPeers.Add(payload.What.NetPeer) == false)
                {
                // Log that this failed
                return;
                }
            }
```

Reply:
I figured that the locking of the whole function is more thread-safe. I'll monitor it's behavior and optimize if any issues occur.